### ISSUE TITLE
Ubuntu 14.04 installation problem.

### ISSUE THREAD

**Issue by sur08** [Lymphatus/CaesiumCLT#10]
I am trying to install in ubuntu 14.04. I successfully installed zopfli but I don't know why but still I am getting this error  `Cannot find libzopflipng.`

**Comment by Lymphatus** [comment]
The package you installed is just the zopfli executable. You need to install the actual PNG library.
Try with:
```
git clone https://github.com/google/zopfli.git
cd zopfli
make libzopflipng
cp libzopflipng.so.1.0.0 /usr/lib
ln -s libzopflipng.so.1.0.0 /usr/lib/libzopflipng.so.1
cp -r src/zopflipng /usr/lib
cd ..
```

**Comment by sur08** [comment]
Hey @Lymphatus thank you :) No luck with also :(

**Comment by Lymphatus** [comment]
All the instructions I gave you went ok?

**Comment by sur08** [comment]
Yes all went okay but getting same error again :(

**Comment by Lymphatus** [comment]
Ok, It took a while to find the issue but I finally figured it out. I will push a fix in the next few hours.

**Comment by pcjonathan** [comment]
I'm having issues with installing on Ubuntu too. (First tried 14.04 but now trying it on 16.04 since that's the current LTS release).

I got the `Cannot find libzopflipng.` error too. Symlinking the libzopflipng.so file fixed that:
```
cd /usr/lib
ln -s libzopflipng.so.1.0.0 libzopflipng.so
```

Then I got the following:
```
checking zopflipng/zopflipng_lib.h usability... no
checking zopflipng/zopflipng_lib.h presence... no
checking for zopflipng/zopflipng_lib.h... no
Cannot find zopflipng/zopflipng_lib.h.
```
Copying the src files (I'd imagine you'd just need the header file but I'm lazy) fixed that.

```cp -r /usr/lib/zopflipng/ /usr/include```

Now `./configure` completes but now I get the following error when trying to make it.

```
gcc -Wall -D_FILE_OFFSET_BITS=64 -std=c99 -g -O2   -o caesiumclt caesiumclt-main.o caesiumclt-jpeg.o caesiumclt-compresshelper.o caesiumclt-utils.o caesiumclt-png.o caesiumclt-lodepng.o caesiumclt-error.o -lm -lzopflipng -lturbojpeg -ljpeg
caesiumclt-jpeg.o: In function `cclt_jpeg_optimize':
/root/CaesiumCLT/src/jpeg.c:132: undefined reference to `jcopy_markers_execute'
/root/CaesiumCLT/src/jpeg.c:136: undefined reference to `jcopy_markers_execute'
collect2: error: ld returned 1 exit status
Makefile:327: recipe for target 'caesiumclt' failed
make[1]: *** [caesiumclt] Error 1
make[1]: Leaving directory '/root/CaesiumCLT/src'
Makefile:376: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by Lymphatus** [comment]
@pcjonathan is absolutely correct. Zopflipng has no official Ubuntu package and you need both headers and dynamic libraries on your system. The `undefined reference to jcopy_markers_execute` is something I recently discovered, since I'm working on OS X and behaves a little differently. I'm going to wrap all it up and write a detailed guide about compiling on Ubuntu (so should be for every Linux system, basically) fixing the latter error in the meanwhile. Goin' to work, then.

**Comment by Lymphatus** [comment]
Ok, I finally came up with a "clean" solution. I tested it in a fresh Ubuntu 16.04 VM (64 bit)

### Prerequisites
- Make sure you have the necessary build tools
       
        $ sudo apt-get install libtool autoconf git nasm





### Build and install mozjpeg
```
$ git clone https://github.com/mozilla/mozjpeg
$ cd mozjpeg/
$ autoreconf -fiv
$ mkdir build && cd build
$ ../configure
$ make && sudo make install
```
Note that you can also use the standard libjpeg-turbo from the official repository, but performance is very different.

### Build and install libzopflipng
```
$ git clone https://github.com/google/zopfli.git
$ cd zopfli
$ make libzopflipng
$ sudo cp libzopflipng.so.1.0.0 /usr/lib
$ sudo ln -s libzopflipng.so.1.0.0 /usr/lib/libzopflipng.so
$ sudo ln -s libzopflipng.so.1.0.0 /usr/lib/libzopflipng.so.1
$ sudo mkdir /usr/include/zopflipng
$ sudo cp src/zopflipng/zopflipng_lib.h /usr/include/zopflipng
```

Basically we build the library, then create some symlinks and copy the header into the include folder.

### Then you can finally build CaesiumCLT
```
$ git clone https://github.com/Lymphatus/CaesiumCLT
$ cd CaesiumCLT/
$ autoreconf -fiv
$ mkdir build && cd build
$ ../configure CFLAGS="-L/opt/mozjpeg/lib64 -I/opt/mozjpeg/include"
$ make && sudo make install
```

I feel there're A LOT of unnecessary and boring thing. I will eventually come up with a cleaner solution in the future.

**Comment by pcjonathan** [comment]
Thanks! I got it installed and have successfully optimised lots of files (in a batch script with WSL). However, it crashes whenever I try to optimise a folder with the following error:

`Segmentation fault (core dumped)`

**Comment by kashu** [comment]
OS: Xubuntu 14.04.5 LTS 64-bit
It works well when I followed the instructions above.
Thank you!

**Comment by kashu** [comment]
> it crashes whenever I try to optimise a folder with the following error:
Segmentation fault (core dumped)

Same here.

**Comment by Lymphatus** [comment]
Closing this issue and opening a new one for the folder crash.
